posydon/popsyn/GRB.py

```diff
@@ -0,0 +1,119 @@
+__author__ = ['Simone Bavera <Simone.Bavera@unige.ch>']
+
+import warnings
+import numpy as np
+from posydon.utils.constants import Msun, clight
+
+GRB_PROPERTIES = ['GRB1', 'S1_eta', 'S1_f_beaming', 'S1_E_GRB', 'S1_E_GRB_iso', 'S1_L_GRB_iso',
+                  'GRB2', 'S2_eta', 'S2_f_beaming', 'S2_E_GRB', 'S2_E_GRB_iso', 'S2_L_GRB_iso']
+
+def get_GRB_properties(df, GRB_efficiency, GRB_beaming, E_GRB_iso_min=0.):
+    """Compute GRB properties for a given binary population."""
+    
+    # check if radiated disk mass is avaialble to compute GRB properties
+    if ('S1_m_disk_radiated' not in df or
+        'S2_m_disk_radiated' not in df):
+        raise ValueError('m_disk_radiated not found in the dataframe!')
+    
+    # reminder the user about the threshold cut
+    if E_GRB_iso_min != 0.:
+        warnings.warn("We only consider GRBs with isotropic equivalent "
+                        f"energy larger than {E_GRB_iso_min} erg!")
+        
+    # define some methods
+    
+    def compute_eta(i, sel):
+        """Compute GRB energy conversion efficiency factor."""
+        if (not isinstance(GRB_efficiency, float) and
+            GRB_efficiency >= 0. and GRB_efficiency <= 1.):
+            raise ValueError(f'GRB efficiency {GRB_efficiency} '
+                             'must be float value in [0,1] range!')
+        # preset all efficiency factors to NaN and set them
+        # only for the selected binaries emitting a GRB
+        df[f'S{i}_eta'] = [np.nan] * df.shape[0]
+        df.loc[sel,f'S{i}_eta'] = GRB_efficiency
+    
+    def compute_beaming_factor(i, sel):
+        """Compute GRB beaming factor."""
+        # TODO: check out Github copilot suggestions: 
+        # 'Fong+15', 'Ghirlanda+16', 'Lamb+19'?
+        
+        df[f'S{i}_f_beaming'] = [np.nan] * df.shape[0]
+        if isinstance(GRB_beaming, float):
+            df.loc[sel, f'S{i}_f_beaming'] = GRB_beaming
+        elif GRB_beaming == 'Goldstein+15':
+            n = df.loc[sel,f'S{i}_f_beaming'].shape[0]
+            # generate a opening angle from a lognormal distribution 
+            # as in Table 3 of https://arxiv.org/pdf/1512.04464.pdf
+            theta = 10**(np.random.normal(0.77,0.37,n))/180.*np.pi # radian
+            # compute the solid angle, factor 2 to accounts for the two emispheres
+            solid_angle = 4*np.pi*(1.-np.cos(theta)) 
+            df.loc[sel,f'S{i}_f_beaming'] = solid_angle/(4*np.pi)
+        elif GRB_beaming == 'Lloyd-Ronning+19':
+            if z is None:
+                raise ValueError('Redshift not specified!')
+            else:
+                z = df.loc[sel,'z_formation']
+            n = df.loc[sel,f'S{i}_f_beaming'].shape[0]
+            # generate a opening angle from a lognormal distribution
+            # as in Table 3 of https://arxiv.org/pdf/1512.04464.pdf
+            theta =10**(np.random.normal(0.77*(1+z)**(-0.75),0.37,n))/180.*np.pi # radian
```

Comment:
I'd put the scaling with z outside the original calculation for theta, because they say in the paper hat theta ~ (1+z)^-0.75; otherwise we would need to think about a z-dependence on sigma.

Reply:
I will comment out the option altogether, with a note to fix it.